src/System.Security.AccessControl/tests/Ace.Tests.cs

```diff
@@ -0,0 +1,482 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Principal;
+using Xunit;
+
+namespace System.Security.AccessControl.Tests
+{
+    public abstract class GenericAce_Tests
+    {
+        public void GenericAce_VerifyAces(GenericAce expectedAce, GenericAce resultAce)
+        {
+            Assert.Equal(expectedAce.AceType, resultAce.AceType);
+            Assert.Equal(expectedAce.AceFlags, resultAce.AceFlags);
+            Assert.Equal(expectedAce.AuditFlags, resultAce.AuditFlags);
+            Assert.Equal(expectedAce.BinaryLength, resultAce.BinaryLength);
+            Assert.Equal(expectedAce.InheritanceFlags, resultAce.InheritanceFlags);
+            Assert.Equal(expectedAce.IsInherited, resultAce.IsInherited);
+            Assert.Equal(expectedAce.PropagationFlags, resultAce.PropagationFlags);
+        }
+
+        public void GenericAce_VerifyBinaryForms(byte[] expectedBinaryForm, byte[] resultBinaryForm, int resultOffset)
+        {
+            Assert.Equal(expectedBinaryForm, resultBinaryForm.Skip(resultOffset));
+        }
+    }
+
+    public abstract class KnownAce_Tests : GenericAce_Tests
+    {
+
+    }
+
+    public class CustomAce_Tests : GenericAce_Tests
+    {
+        private static object[] CustomAce_CreateTestData(int intType, int intFlags, int opaqueLength, int offset)
+        {
+            byte[] opaque = new byte[opaqueLength];
+            AceType type = (AceType)intType;
+            AceFlags flags = (AceFlags)intFlags;
+
+            CustomAce ace = new CustomAce(type, flags, opaque);
+            Assert.Equal(type, ace.AceType);
+            Assert.Equal(flags, ace.AceFlags);
+            Assert.Equal(opaque, ace.GetOpaque());
+
+            byte[] binaryForm = new byte[ace.BinaryLength + offset];
+            binaryForm[offset + 0] = (byte)type;
+            binaryForm[offset + 1] = (byte)flags;
+            binaryForm[offset + 2] = (byte)(ace.BinaryLength >> 0);
+            binaryForm[offset + 3] = (byte)(ace.BinaryLength >> 8);
+            opaque.CopyTo(binaryForm, 4 + offset);
+
+            return new object[] { ace, binaryForm, offset };
+        }
+
+        public static IEnumerable<object[]> CustomAce_TestObjects()
+        {
+            yield return CustomAce_CreateTestData(19, 0, 4, 0);
+            yield return CustomAce_CreateTestData(18, 1, 8, 0);
+            yield return CustomAce_CreateTestData(17, 1, 4, 0);
+        }
+
+        [Fact]
+        public void CustomAce_Constructor_Invalid()
+        {
+            Assert.Throws<ArgumentOutOfRangeException>(() => new CustomAce((AceType)16, (AceFlags)15, new byte[1]));
+            Assert.Throws<ArgumentOutOfRangeException>(() => new CustomAce((AceType)19, (AceFlags)1, new byte[1]));
+            Assert.Throws<ArgumentOutOfRangeException>(() => new CustomAce((AceType)0, (AceFlags)2, new byte[4]));
```

Comment:
Why are any of these invalid?  Is there value in further checking the parameter name of the AOORE?